### PR TITLE
Add the internal access modifier for types, operations, and functions

### LINKED ISSUE #259: Access modifiers for callables and types
**Is your feature request related to a problem? Please describe.**
When developing Q# libraries, it is common to introduce small functions and operations that are not intended for direct use, but which are helpful with partial application. At the moment, these functions and operations are given names starting with `_` or ending with `Impl`, but that makes it easy to accidentally take a dependency on these functions and operations from other namespaces or compilation units.

**Describe the solution you'd like**
I would suggest the inclusion of an accessibility modifier attribute (or possibly multiple, depending on the level of granularity needed), paralleling the use of special-purpose keywords in other languages. In particular,
```Q#
namespace Microsoft.Quantum.Core {
 
    @Attribute()
    newtype Private = Unit;

}
```

In code that calls a `@Private()` function or operation from another namespace or compilation unit, a warning could be raised indicating that about a possible accidental dependency on internal implementation details.

The new accessibility modifier attributes could also be used to simplify documentation generation logic, such that no documentation is generated for `@Private()` callables rather than depending on name-matching (paralleling the new `@Test("…")` attribute).

**Describe alternatives you've considered**
- The current approach could continue to be used, in which the leading underscore convention is given special meaning by the compiler and documentation generation tool. This suffers the same disadvantages as relying on `*Test` names for unit testing.
- New keywords (e.g.: `private`) could be introduced instead of new attributes.

### PR DESCRIPTION
**Pull request by marshallsa**
Resolves #259.

TODO:
- [x] Name mangling for imported internal declarations (#370)
- [x] Support for re-using internal names in more than one reference (#376)

### CODE REVIEW

File: src/QsCompiler/CompilationManager/CompilationUnit.cs [hunk 1]
```diff
@@ -143,6 +153,8 @@ public References(ImmutableDictionary<NonNullable<string>, Headers> refs, Action
     /// </summary>
     public class CompilationUnit : IReaderWriterLock, IDisposable
     {
+        internal static readonly NameDecorator ReferenceDecorator = new NameDecorator("QsReference");
```
Comment (bettinaheim) [inline src/QsCompiler/CompilationManager/CompilationUnit.cs]: Looks like this one can actually be private.

File: src/QsCompiler/CompilationManager/EditorSupport/CodeActions.cs [hunk 2]
```diff
@@ -38,24 +39,33 @@ private static WorkspaceEdit GetWorkspaceEdit(this FileContentManager file, para
...
+        /// 
+        /// Returns the name of the identifier as an out parameter if an unqualified symbol exists at that location.
         /// </summary>
         private static IEnumerable<NonNullable<string>> NamespaceSuggestionsForIdAtPosition
             (this FileContentManager file, Position pos, CompilationUnit compilation, out string idName)
         {
             var variables = file?.TryGetQsSymbolInfo(pos, true, out CodeFragment _)?.UsedVariables;
             idName = variables != null && variables.Any() ? variables.Single().Symbol.AsDeclarationName(null) : null;
-            return idName != null && compilation != null
+            var nsName = file.TryGetNamespaceAt(pos);
```
Comment (bettinaheim) [inline src/QsCompiler/CompilationManager/EditorSupport/CodeActions.cs]: This additional query for the namespace is no longer needed (it was needed with private I believe). I'd suggest to remove it again for performance reasons.

File: src/QsCompiler/CompilationManager/EditorSupport/CodeActions.cs [hunk 3]
```diff
@@ -64,7 +74,8 @@ private static IEnumerable<NonNullable<string>> NamespaceSuggestionsForTypeAtPos
             typeName = types != null && types.Any() &&
                 types.Single().Type is QsTypeKind<QsType, QsSymbol, QsSymbol, Characteristics>.UserDefinedType udt
                 ? udt.Item.Symbol.AsDeclarationName(null) : null;
-            return typeName != null && compilation != null
+            var nsName = file.TryGetNamespaceAt(pos);
```
Comment (bettinaheim) [inline src/QsCompiler/CompilationManager/EditorSupport/CodeActions.cs]: same here.

**Comment by bettinaheim** [comment]
Two minor comments, but other than that looks good! Well done, @marshallsa!

File: src/QsCompiler/CompilationManager/CompilationUnit.cs (hunk 1 above)
Comment (marshallsa) [reply src/QsCompiler/CompilationManager/CompilationUnit.cs]: It's used by the linking tests.

Comment (bettinaheim) [reply src/QsCompiler/CompilationManager/CompilationUnit.cs]: Hu, I could have sworn I built it. But I'll take your word for it.

**Pull request merged**